Various fixes for transfer ownership

LEDfan opened the pull request:
First of all this adds the `--move` option, this will move the data into the root of the target user, but only if this user doesn't have any files. This can be used as a workaround for renaming users (which is required in one setup when the UID of ldap changes).

The three other commits are improvements to the transfer-ownership command. During my tests on a real instance, there was some issue with group folders and also the check to prevent to move to users who never logged in didn't work.

The last commit adds a catch for PHP errors. Sometimes the owner of a share was null, and this would stop the transfer script.

I guess it's okay to have this one PR, if not let me now and I will split it up.

LEDfan committed:
Fix transfer-ownership for Group Folders

Because Group folders don't have owners, use the same system to check
the permissions of the node.

Signed-off-by: Tobia De Koninck <LEDfan@users.noreply.github.com>

LEDfan committed:
Fix tests

Signed-off-by: Tobia De Koninck <LEDfan@users.noreply.github.com>

skjnldsv commented:
Please rebase @LEDfan :)

blizzz commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -144,6 +154,11 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 
+		if ($input->hasArgument('move') && (!$view->is_dir($this->finalTarget) || count($view->getDirectoryContent($this->finalTarget)) > 0)) {
+			$output->writeln("<error>Destination path does not exists or is not empty</error>");
+			return 1;
```
return a different error code

blizzz reviewed (commented):
looks ok, but two nitpicks. didn't test yet.

blizzz commented:
> During my tests on a real instance, there was some issue with group folders and also the check to prevent to move to users who never logged in didn't work.

Can you elaborate on that?

We shouldn't have direct references to code from other apps in here. Maybe we need a fix in the app, or a more general rule in this code?

blizzz commented:
(ah, btw, there is no notification issued with only new commits, but no comment)

ChristophWurst commented:
@LEDfan could you have another look, please? Since your PR we've extracted the code into a reusable service #18025 because it's not only used by the CLI command in Nextcloud 18. Thus some of the changes have to be applied there.

wiswedel commented:
> @LEDfan could you have another look, please? Since your PR we've extracted the code into a reusable service #18025 because it's not only used by the CLI command in Nextcloud 18. Thus some of the changes have to be applied there.

Could someone please speed this up so we can start digging into https://github.com/nextcloud/guests/issues/223 (which I need for a customer)?

LEDfan committed:
Add option to transfer-ownership to move data

This will move the home folder of own user to another user. Only allowed
if that other user's home folder is empty.
Can be used as workaround to rename users.

Signed-off-by: Tobia De Koninck <LEDfan@users.noreply.github.com>

LEDfan committed:
Prevent transferring data to user which never loggedin

Signed-off-by: Tobia De Koninck <LEDfan@users.noreply.github.com>

LEDfan committed:
Catch \Error in Transfer::restoreShares

This makes the command more fault tolerant. An \Error can happen when
e.g. the owner of a share is null.
If we don't catch this, the restore process will stop in an unknown
state.

Signed-off-by: Tobia De Koninck <LEDfan@users.noreply.github.com>

LEDfan commented:
@ChristophWurst @wiswedel I applied the changes to the new code and quickly tested it. I'm currently doing my master thesis, so don't have much time to thoroughly test it.

>>   During my tests on a real instance, there was some issue with group folders and also the check to prevent to move to users who never logged in didn't work.

> Can you elaborate on that?
> We shouldn't have direct references to code from other apps in here. Maybe we need a fix in the app, or a more general rule in this code?

I understand. IIRC the problem with the group folders was something like: 
 - assume there is a group folder `GF_test` shared to the `everyone` group. This group has some user s user1-user4
 - user2 creates a folder in `GF_test`
 - user2 shares this folder (in the GF) again with user3
 - try do transfer from user3 to user4

On the instance I was speaking of, this throws exceptions and fails the transfer. 
I just tried to reproduce this on the current master. However, the shared folder with user3 does not appear in the home folder of user3. In the section `Shared with you` there is also no folder, but there is the text `1 folder`.  The transfer command also now works without the change.

![Shared with you - Nextcloud - Chromium_1283](https://user-images.githubusercontent.com/2996275/71713170-dcf6cf80-2e08-11ea-9e36-af347f019d4f.png)

Maybe something has changed in NC18 regarding to re-sharing folders inside groupfolders?
So, I guess it may the safest option to remove the commit related to groupfolders.

wiswedel commented:
>  So, I guess it may the safest option to remove the commit related to groupfolders.

I agree. Group folders adds a hell lot of complexity to the transfer operation. I vote for skipping group folder shares for the time being and mark that as an open issue. Better this way covering all "normal" shares than having a rushed implementation with hidden non-tested corner cases

ChristophWurst commented on apps/files/lib/Command/TransferOwnership.php:
```diff
@@ -99,7 +104,8 @@ protected function execute(InputInterface $input, OutputInterface $output) {
 				ltrim($input->getOption('path'), '/'),
-				$output
+				$output,
+				$input->hasArgument('move')
```
IIRC this always returns true as the parameter exists. You can use `getOption('move') !== null`

ChristophWurst reviewed (approved):
Code looks good!

rullzer commented:
> > So, I guess it may the safest option to remove the commit related to groupfolders.
> 
> I agree. Group folders adds a hell lot of complexity to the transfer operation. I vote for skipping group folder shares for the time being and mark that as an open issue. Better this way covering all "normal" shares than having a rushed implementation with hidden non-tested corner cases

Group folders should not fall in the transfer ownership transfer at all. Since the user doesn't own them ;)

rullzer commented on apps/files/lib/Service/OwnershipTransferService.php:
```diff
@@ -99,6 +107,11 @@ public function transfer(IUser $sourceUser,
 		}
 
+		if ($move && (!$view->is_dir($finalTarget) || count($view->getDirectoryContent($finalTarget)) > 0)) {
+			throw new TransferOwnershipException("Destination path does not exists or is not empty", 1);
```
Wouldn't this fail in 99% of the cases?
As the example files are put into place?

ChristophWurst commented:
No update in a long time and conflicts badly. Let's do this from scratch if it's still relevant. If I'm not mistaken we added some of the changes in other PRs.
